system/jhtree/bloom.cpp

```diff
@@ -75,32 +77,167 @@ bool BloomFilter::test(hash64_t hash) const
     return true;
 }
 
-BloomBuilder::BloomBuilder(unsigned _maxHashes) : maxHashes(_maxHashes)
+IndexBloomFilter::IndexBloomFilter(unsigned _numHashes, unsigned _tableSize, byte *_table, __uint64 _fields)
+: BloomFilter(_numHashes, _tableSize, _table), fields(_fields)
+{}
+
+int IndexBloomFilter::compare(CInterface *const *_a, CInterface *const *_b)
+{
+    const IndexBloomFilter *a = static_cast<IndexBloomFilter *>(*_a);
+    const IndexBloomFilter *b = static_cast<IndexBloomFilter *>(*_b);
+    return a->fields - b->fields;
+}
+
+bool IndexBloomFilter::reject(const SegMonitorList &segs) const
+{
+    hash64_t hashval = HASH64_INIT;
+    return getBloomHash(fields, segs, hashval) && !test(hashval);
+}
+
+extern bool getBloomHash(__int64 fields, const SegMonitorList &segs, hash64_t &hashval)
+{
+    while (fields)
+    {
+        unsigned f = ffsll(fields)-1;    // extract lowest 1 bit
```

Comment:
This will not compile on windows.  Need to use countTrailingUnsetBits() or similar.